docs/the-new-architecture/pillars-turbomodule.md

```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+  }
+  mavenCentral()
+  google()
+}
+
+dependencies {
+  implementation 'com.facebook.react:react-native:+'
+}
+
+react {
+    jsRootDir = file("../js/")
+    libraryName = "RTNCalculator"
+    codegenJavaPackageName = "com.RTNCalculator"
+}
+```
+
+Of interest in the `build.gradle` file:
+
+- The `defaultConfig` block, within the `android` block, adds a `buildConfigField` to enable the New Architecture.
+- The `react` block configures the CodeGen process. For Android, we need to specify:
+  - the `jsRootDir`, which contains the relative path to the JavaScript specs
+  - the `libraryName` we will use to link the library in the app.
+  - the `codegenJavaPackageName` which corresponds to the name of the Java package we will use for the code generated by **CodeGen**.
+
+## 4. Native Code
+
+For the final step in getting your TurboModule ready to go, you'll need to write some native code to connect the JavaScript side to the native platforms. This process requires two main steps:
+
+- Run **CodeGen** to see what it generates.
+- Write your native code, implementing the generated interfaces.
+
+When developing a React Native app that uses a TurboMOdule, it is responsibility of the app to actually generate the code using **CodeGen**. However, when developing a TurboModule as a library, we need to reference the generated code, and it is therefore useful to see what the app will generate.
+
+As first step for both iOS and Android, this guide shows how to execute manually the scripts used by **CodeGen** to generate the required code. Further information on **CodeGen** can be found [here](/docs/pillars-codegen.md)
+
+:::caution
+The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.
+:::
+
+### iOS
+
+#### Generate the code - iOS
+
+To run Codegen for the iOS platform, we need to open a terminal and run the following command:
+
+```sh title="Running CodeGen for iOS"
+cd MyApp
+yarn add ../RTNCalculator
+cd ..
+node MyApp/node_modules/react-native/scripts/generate-artifacts.js \
+  --path MyApp/ \
+  --outputPath RTNCalculator/generated/
+```
+
+This script first adds the `RTNCalculator` module to the app with `yarn add`. Then, it invokes Codegen via the `generate-artifacts.js` script.
+
+The `--path` option specifies the path to the app, while the `--outputPath` option tells the script where to output the generated code.
+
+The output of this process is the following folder structure:
+
+```sh
+generated
+└── build
+    └── generated
+        └── ios
+            ├── FBReactNativeSpec
+            │   ├── FBReactNativeSpec-generated.mm
+            │   └── FBReactNativeSpec.h
+            ├── RCTThirdPartyFabricComponentsProvider.h
+            ├── RCTThirdPartyFabricComponentsProvider.mm
+            ├── RTNCalculatorSpec
```

Comment:
Something that was surprising to me: after I generated this the first time, I realized I'd made a mistake in naming the module (RN instead of RTN), so changed the code and ran this again.

Instead of replacing the generated folder contents completely, it added the new `RTNCalculatorSpec` _alongside_ the `RNCalculatorSpec` that had been removed. If I totally wiped the generated files and re-generated them, `RNCalculatorSpec` was gone. Is this expected behavior?

Reply:
> Instead of replacing the generated folder contents completely

IIRC the generate-artifacts script is not responsible of wiping the output folder, but just of reading the input and producing an output in the specified folder.

On Android: the cleanup is done by the Gradle Plugin. On iOS we should have a similar mechanism (cc @cipolleschi)

Reply:
The cleanup mechanism on iOS is there on `main` but not in 0.68 nor in 0.69. :( 
At the moment, the folder has to be removed manually, unfortunately.

Reply:
Should the docs include a warning in case anyone runs into this?

Reply:
> @cortinico, we wipe the `build` folder every time we run the CodeGen on Android?

Answering for future reference: yes we do wipe the generated folder every time the Codegen needs to run:

https://github.com/facebook/react-native/blob/8a33b75f55ba429e9e7a897179557aa8bb49ab35/packages/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/tasks/GenerateCodegenSchemaTask.kt#L49-L52

This is to prevent stale files from being around.

> Should the docs include a warning in case anyone runs into this?

IMHO not. I would not mention this as this is how the whole Android/Gradle environment works (things are not re-executed unless needed). If there is a issue, is most likely a bug that needs to be addressed on our end 👍